staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go

```diff
@@ -0,0 +1,453 @@
+)
+
+const (
+	azureTimeFormat = "\"2006-01-02 15:04:05.99999\""
+	azureAuthority  = "https://login.microsoftonline.com"
+	uuidFormat      = "[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[8|9|aA|bB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}"
+	azureTokenKey   = "azureTokenKey"
+
+	cfgClientID     = "client-id"
+	cfgTenantID     = "tenant-id"
+	cfgAudience     = "audience"
+	cfgAccessToken  = "access-token"
+	cfgRefreshToken = "refresh-token"
+	cfgTokenType    = "token-type"
+	cfgExpiresIn    = "expires-in"
+	cfgExpiresOn    = "expires-on"
+)
+
+func init() {
+	if err := restclient.RegisterAuthProviderPlugin("azure", newAzureAuthProvider); err != nil {
+		glog.Fatalf("Failed to register azure auth plugin: %v", err)
+	}
+}
+
+var cache = newAzureTokenCache()
+
+type azureTokenCache struct {
+	lock  sync.Mutex
+	cache map[string]*azureToken
+}
+
+func newAzureTokenCache() *azureTokenCache {
+	return &azureTokenCache{cache: make(map[string]*azureToken)}
+}
+
+func (c *azureTokenCache) getToken(tokenKey string) *azureToken {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.cache[tokenKey]
+}
+
+func (c *azureTokenCache) setToken(tokenKey string, token *azureToken) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.cache[tokenKey] = token
+}
+
+func newAzureAuthProvider(_ string, cfg map[string]string, persister restclient.AuthProviderConfigPersister) (restclient.AuthProvider, error) {
+	var ts tokenSource
+
+	clientID := cfg[cfgClientID]
+	tenantID := cfg[cfgTenantID]
+	audience := cfg[cfgAudience]
+	if clientID != "" && tenantID != "" && audience != "" {
+		ts = newAzureTokenSourceDeviceCode(clientID, tenantID, audience)
+	} else {
+		azFilename, err := azureTokensFile()
+		if err != nil {
+			return nil, err
+		}
+		ts, err = newAzureTokenSourceFromFile(azFilename)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	cacheSource := newAzureTokenSourceFromCache(ts, cache, cfg, persister)
+
+	return &azureAuthProvider{
+		tokenSource: cacheSource,
+	}, nil
+}
+
+type azureAuthProvider struct {
+	tokenSource tokenSource
+}
+
+func (p *azureAuthProvider) Login() error {
+	return errors.New("not yet implemented")
+}
+
+func (p *azureAuthProvider) WrapTransport(rt http.RoundTripper) http.RoundTripper {
+	return &azureRoundTripper{
+		tokenSource:  p.tokenSource,
+		roundTripper: rt,
+	}
+}
+
+type azureRoundTripper struct {
+	tokenSource  tokenSource
+	roundTripper http.RoundTripper
+}
+
+func (r *azureRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	token, err := r.tokenSource.Token()
+	if err != nil {
+		glog.Errorf("Failed to acquire a token: %v", err)
+		return nil, err
+	}
+
+	// clone the request in order to avoid modifying the headers of the original request
+	req2 := new(http.Request)
+	*req2 = *req
+	req2.Header = make(http.Header, len(req.Header))
+	for k, s := range req.Header {
+		req2.Header[k] = append([]string(nil), s...)
+	}
+
+	req2.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token.token.AccessToken))
```

Comment:
This is right the `id_token` issued by AAD cannot be used to verity the access, but as I said the access token contains the same claims. 

It is a bit strange how the `appiserver` validates the `audience`. The identity provider will request consent to a user for a specific audience when issuing the token. In this case should be a dedicated audience for ` kubernetes apiserver` which should be registered in the identity provider as resource server.

The issue is in CorsOS's `go-oidc` package which is used by the oidc plugin of apiserver. It property verifies the signature, expiration but it does the following assumption on audience. 

https://github.com/coreos/go-oidc/blob/master/oidc/verification.go

Code snippet from `VerifyClaims` function
```Go
	// aud REQUIRED. Audience(s) that this ID Token is intended for.
	// It MUST contain the OAuth 2.0 client_id of the Relying Party as an audience value.
	// It MAY also contain identifiers for other audiences. In the general case, the aud
	// value is an array of case sensitive strings. In the common special case when there
	// is one audience, the aud value MAY be a single case sensitive string.
	if aud, ok, err := claims.StringClaim("aud"); err == nil && ok {
		if aud != clientID {
			return fmt.Errorf("invalid claims, 'aud' claim and 'client_id' do not match, aud=%s, client_id=%s", aud, clientID)
		}
	} else if aud, ok, err := claims.StringsClaim("aud"); err == nil && ok {
		if !containsString(clientID, aud) {
			return fmt.Errorf("invalid claims, cannot find 'client_id' in 'aud' claim, aud=%v, client_id=%s", aud, clientID)
		}
	} else {
		return errors.New("invalid claim value: 'aud' is required, and should be either string or string array")
	}
```